apiserver/tools.go

```diff
@@ -20,20 +20,121 @@ import (
 	"github.com/juju/juju/apiserver/common"
 	"github.com/juju/juju/apiserver/httpcontext"
 	internalhttp "github.com/juju/juju/apiserver/internal/http"
+	coreagentbinary "github.com/juju/juju/core/agentbinary"
+	coreerrors "github.com/juju/juju/core/errors"
 	coremodel "github.com/juju/juju/core/model"
 	"github.com/juju/juju/core/objectstore"
 	"github.com/juju/juju/core/semversion"
+	agentbinaryerrors "github.com/juju/juju/domain/agentbinary/errors"
 	"github.com/juju/juju/environs"
 	"github.com/juju/juju/environs/simplestreams"
 	envtools "github.com/juju/juju/environs/tools"
+	internalerrors "github.com/juju/juju/internal/errors"
 	jujuhttp "github.com/juju/juju/internal/http"
+	"github.com/juju/juju/internal/services"
 	"github.com/juju/juju/internal/tools"
 	"github.com/juju/juju/rpc/params"
 	"github.com/juju/juju/state"
 	"github.com/juju/juju/state/binarystorage"
 	"github.com/juju/juju/state/stateenvirons"
 )
 
+// AgentBinaryStore is an interface that provides the ability to store new agent
+// binaries into a store for the controller or model.
+type AgentBinaryStore interface {
+	// AddAgentBinaryWithSHA256 adds a new agent binary to the object store and
+	// saves its metadata to the database.
+	// The following errors can be returned:
+	// - [coreerrors.NotSupported] if the architecture is not supported.
+	// - [agentbinaryerrors.AlreadyExists] if an agent binary already exists for
+	// this version and architecture.
+	// - [agentbinaryerrors.ObjectNotFound] if there was a problem referencing
+	// the agent binary metadata with the previously saved binary object.
+	// This error should be considered an internal problem. It is discussed here
+	// to make the caller aware of future problems.
+	// - [coreerrors.NotValid] if the agent version is not valid or the SHA256
```

Comment:
I think a sha mismatch deserves its own error type. It's a distinct class of problem (even a potential security issue) compared to an invalid version string.

Reply:
Done & Done

Reply:
Done & Done